METRICS.md

```diff
@@ -0,0 +1,155 @@
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+-->
+
+# Apache NiFi - MiNiFi - C++ Metrics Readme.
+
+
+This readme defines the metrics published by Apache NiFi. All options defined are located in minifi.properties.
+
+## Table of Contents
+
+- [Description](#description)
+- [Configuration](#configuration)
+- [Metrics](#metrics)
+
+## Description
+
+Apache NiFi MiNiFi C++ can communicate metrics about the agent's status, that can be a system level or component level metric.
+These metrics are exposed through the agent implemented metric publishers that can be configured in the minifi.properties.
+Aside from the publisher exposed metrics, metrics are also sent through C2 protocol of which there is more information in the
+[C2 documentation](C2.md#metrics).
+
+## Configuration
+
+To configure the a metrics publisher first we have to set which publisher class should be used:
+
+	# in minifi.properties
+
+	nifi.metrics.publisher.class=PrometheusMetricsPublisher
+
+Currently PrometheusMetricsPublisher is the only available publisher in MiNiFi C++ which publishes metrics to a Prometheus server.
+To use the publisher a port should also be configured where the metrics will be available to be scraped through:
+
+	# in minifi.properties
+
+	nifi.metrics.publisher.PrometheusMetricsPublisher.port=9936
+
+The last option defines which metric classes should be exposed through the metrics publisher in configured with a comma separated value:
+
+	# in minifi.properties
+
+	nifi.metrics.publisher.metrics=QueueMetrics,RepositoryMetrics,GetFileMetrics,DeviceInfoNode,FlowInformation
+
+## Metrics
+
+The following section defines the currently available metrics to be published by the MiNiFi C++ agent.
+
+NOTE: In Prometheus all metrics are extended with a `minifi_` prefix to mark the domain of the metric. For example the `connection_name` metric is published as `minifi_connection_name` in Prometheus.
+
+### QueueMetrics
+
+QueueMetrics is a system level metric that reports queue metrics for every connection in the flow.
+
+| Metric name          | Labels                                         | Description                                |
+|----------------------|------------------------------------------------|--------------------------------------------|
+| queue_data_size      | metric_class, connection_uuid, connection_name | Max queue size to apply back pressure      |
+| queue_data_size_max  | metric_class, connection_uuid, connection_name | Max queue data size to apply back pressure |
+| queue_size           | metric_class, connection_uuid, connection_name | Current queue size                         |
+| queue_size_max       | metric_class, connection_uuid, connection_name | Current queue data size                    |
```

Comment:
I've found this guide to metric naming: https://prometheus.io/docs/practices/naming/

By the way I think the descriptions are not in the correct order here.

Reply:
I realize it would be hard to have metric sink-specific names without some type metadata for metrics, and I wouldn't want to touch the C2 protocol. Let's leave it as is for now, and think about changing the prometheus metrics names later, after some metrics-related refactoring.

Could you check the descriptions though? They are mixed up.

Reply:
Okay, I think we would have more input and would be able to generalize these things a bit easier if we had another monitoring tool support, that would make it easier to see what needs to be more generalized.

I see there is a mixup in the descriptions I'll fix it.